pkg/certificate/certificate.go

```diff
@@ -66,11 +67,55 @@ func ValidateConnectionContext(ctx context.Context, pool *x509.CertPool, fleetUR
 			},
 		},
 	}
-	conn, err := dialer.DialContext(ctx, "tcp", parsed.Host)
+	conn, err := dialer.DialContext(ctx, "tcp", getHostPort(parsed))
 	if err != nil {
 		return ctxerr.Wrap(ctx, err, "dial for validate")
 	}
+	conn.Close()
+
+	return nil
+}
+
+// ValidateClientAuthTLSConnection validates that a TLS connection can be made
+// to the server identified by the target URL (only the host portion is used)
+// by authenticating the client using the provided certificate. The ctx may
+// specify a timeout or deadline for the TLS connection check.
+func ValidateClientAuthTLSConnection(ctx context.Context, cert *tls.Certificate, targetURL string) error {
+	parsed, err := url.Parse(targetURL)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "parse url")
+	}
+
+	dialer := &tls.Dialer{
+		Config: &tls.Config{
+			GetClientCertificate: func(reqInfo *tls.CertificateRequestInfo) (*tls.Certificate, error) {
+				return cert, nil
+			},
+			ServerName:         parsed.Hostname(),
+			ClientSessionCache: tls.NewLRUClientSessionCache(-1),
+		},
+	}
+
+	conn, err := dialer.DialContext(ctx, "tcp", getHostPort(parsed))
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "TLS dial")
+	}
 	defer conn.Close()
 
+	if _, err = conn.Read(make([]byte, 1024)); err != nil {
+		return ctxerr.Wrap(ctx, err, "read from TLS connection")
+	}
 	return nil
 }
+
+func getHostPort(u *url.URL) string {
+	host, port := u.Hostname(), u.Port()
+	if port == "" {
+		if u.Scheme == "https" {
+			port = "443"
+		} else {
+			port = "80"
+		}
+	}
+	return net.JoinHostPort(host, port)
```

Comment:
this is not a request to change any of this, but just because I found it interesting: I was reading `Dial`'s documentation and I found:

```
// For TCP and UDP networks, the address has the form "host:port".
// [..]
// The port must be a literal port number or a service name.
```

so doing `return net.JoinHostPort(u.Hostname(), u.Scheme)` could accomplish the same 🤯

Reply:
Oh that's nice! Yeah I might change it, will simplify things.

Reply:
Though the port _can_ be different than the one associated with the scheme, so it can't all be replaced with that line, just if there's no port specified. Still simpler though.

Reply:
Yeah, I added a comment along with the change. I still think it's nice as you don't have to hard-code port number.